Reduce Application Binary Size

jkmassel opened the pull request:
In #22296, we merged a very large PDF file as an `xcassets` resource.

This increased the binary size by 43.1MB, causing it to be too large for installation or update on cellular networks.

This PR resolves that by:

1. Switching out the PDF file for a JPEG-compressed gradient. If I'd had more time, I would've probably generated the gradient in code instead of as a static image, but that can be left for another day.
2. Running perceptually invisible compression on all PNG assets (which saves another 0.5 MB)

<img width="703" alt="Screenshot 2024-05-27 at 3 18 52 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1123407/e46e134d-c111-4c3d-9e3a-f00d8a7f00d2">

To test:

## Regression Notes
1. Potential unintended areas of impact
- Modifies many icon images – use the left/right comparison on GitHub to note that they're visually indistinguishable.
- Run the first-launch process, note that there's no user-visible difference.

3. What I did to test those areas of impact (or what existing automated tests I relied on)
- I did the items above.

4. What automated tests I added (or what prevented me from doing so)
- No automated tests available for this.
- 
PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

jkmassel committed:
Perceptual compression for PNG assets

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/291">25.0 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jkmassel committed:
Replace huge PDFs with smaller JPGs

guarani commented on WordPress/Resources/AppImages.xcassets/stats-revamp-v2-feature-intro-preview.imageset/Contents.json:
@jkmassel, this resource is unused and can be removed entirely. 
We removed the screen that used it in https://github.com/wordpress-mobile/WordPress-iOS/pull/19917, but the PDF was overlooked.

jkmassel replied:
Great! Addressed in ac030705854d26713fbceaa301726ee45f49d726

guarani reviewed (commented):
Thanks @jkmassel!

I realized the Stats PDF isn't used, so it can be removed. 

I tested the other change to the welcome screen and noticed it looks odd in dark mode:

| this branch | `trunk` |
| - | - |
| <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/32c346b5-1163-447a-9bb3-4cd7345f481b" width="350" /> | <img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/088b8305-4747-4a4b-a9de-ba415181c9c3" width="350" /> |

alpavanoglu commented:
Hey @jkmassel 
> This increased the binary size by 43.1MB

So that I have a better understanding and avoid this kind of thing in the future, how does that pdf file increase the binary size by 43.1 Megabytes when the file itself appear to be 5 Kilobytes? I understand the vector file will be converted but still the number is extremely high indeed. 

If you mean the gradient, shouldn't the trade there lower the size rather than increase it?

![Screenshot 2024-05-28 at 16 21 49](https://github.com/wordpress-mobile/WordPress-iOS/assets/15968946/a433455f-e178-47a1-83eb-78d477772b39)

jkmassel committed:
Use an SVG (and manually apply blur)

Xcode’s SVG handling is interesting – it doesn’t properly handle filters, so the blur isn’t baked into the resulting image.

This is actually a feature, not a bug – the old encoding approach was worst-case scenario for image size – a partially-transparent image with gradients (from the blur) and noise applied overtop; and because Xcode turns the SVG into a PNG for us it would get unexpected large very quickly.

Doing the blur manually takes just as much CPU time as it would if we were processing the SVG within iOS, but results in a much smaller image inside the bundle.

jkmassel committed:
Remove an unused image

jkmassel commented:
@alpavanoglu – I'm not sure I understand?

> how does that pdf file increase the binary size by 43.1 Megabytes when the file itself appear to be 5 Kilobytes

Where did you get 5kb from? The binary size is increased so much because the 7 MB PDF file is converted to a 3x, 2x, and 1x PNG by Xcode during compilation. The resulting `Assets.car` file is duplicated in the App Binary – it's also part of `WordPressShareExtension`, `WordPressDraftActionExtension`,  and`JetpackIntents`, so the cost has to be paid four times.

> If you mean the gradient, shouldn't the trade there lower the size rather than increase it?

Interesting – I suspected #22296 because removing this file reduced the binary size by the 43 MB and we got a notification from Apple on Feb 20 that the app had exceeded the download limit. Perhaps there is another factor at play?

jkmassel commented:
Thanks @guarani – the transparency adds a bunch of complexity – the issue should now be resolved. Unfortunately the savings aren't quite as dramatic as with a JPEG but it's a good start. Brings down the Jetpack binary size from 269.8MB to 210.6MB, so we're much closer to the cellular limit now!

jkmassel committed:
Fix rubocop error

alpavanoglu commented:
@jkmassel 

> Where did you get 5kb from?

That PR added 1 PDF and replaced the other. The pure addition is the small one I referred to there. And the replaced PDF is smaller in size so I reckoned overall I would be reducing the size and not increase it. 

> Perhaps there is another factor at play?

Perhaps. I wanted to clarify to avoid such things in the future.

jkmassel commented:
Addressed in #23381
